### PR TITLE
Implement auth proxy relation and access rules rendering logic

### PR DESCRIPTION
**Pull request by natalian98**
## Summary
This PR:
- adds `auth-proxy` relation interface
- adds a peer relation to store access rules locations for each relation
- adds access rules rendering and removing logic
- modifies config template

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -223,13 +271,134 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                protected_urls=event.protected_urls,
+                allowed_endpoints=event.allowed_endpoints,
+                relation_app_name=event.relation_app_name,
+            )
+
+            if rules:
+                # Push the rules to the container
+                filename = f"{self._oathkeeper_access_rules_path}/access-rules-{event.relation_app_name}-{rule_type}.json"
+                self._container.push(filename, str(rules), make_dirs=True)
+                access_rules_locations.append(filename)
```
Comment (shipperizer) [inline src/charm.py:305]: this should edit the k8s configmap (mounted as a volume) instead of the file

File: src/charm.py [hunk 2]
```diff
@@ -223,13 +271,134 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                url = url.replace("https", "<https|http>")
+
+            if rule_type == "allow":
+                if not allowed_endpoints:
+                    return None
+                for endpoint in allowed_endpoints:
+                    allow_rule = self._rule_template(
+                        rule_id=f"{relation_app_name}:{endpoint}:{url_index}:allow",
+                        url=f"{url}/{endpoint}",
+                        authenticator="noop",
```
Comment (shipperizer) [inline src/charm.py:347]: shouldn't this be `cookie_session`?

Comment (natalian98) [reply src/charm.py:347]: Allow rules use `noop`, so that the request for the given endpoint is allowed without having to log in.
Deny rules use `cookie_session`.

File: src/charm.py [hunk 3]
```diff
@@ -223,13 +271,134 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+
+        # TODO: Render with all relations' locations
+        self._update_oathkeeper_config_with_access_rules_locations(access_rules_locations)
+
+    def _rule_template(
+        self, rule_id: str, url: str, authenticator: str, mutator: str, error_handler: str
+    ) -> Dict:
+        return {
+            "id": rule_id,
+            "match": {"url": url, "methods": ["GET", "POST"]},
```
Comment (shipperizer) [inline src/charm.py:318]:
```suggestion
"match": {"url": url, "methods": ["GET", "POST","OPTION","PUT","PATCH","DELETE"}, 
```

we would need all of them i 'm afraid, OPTIONS for sure, as that is kinda needed for any CORS behaviour

File: src/charm.py [hunk 4]
```diff
@@ -223,13 +271,134 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        if "/" in endpoint:
+                            # about/app must become a group: (about/app)
+                            endpoint = f"({endpoint})"
+                        exclude_endpoints.append(endpoint)
+
+                    # Add | alternation
+                    deny_regex = f"{url}/<(?!{'|'.join(exclude_endpoints)}).*>"
```
Comment (shipperizer) [inline src/charm.py:365]: we would need this in any case, not only if the rule type is "deny"

Comment (natalian98) [reply src/charm.py:365]: For better visibility, each allowed endpoint gets its own rule created instead of a single rule matching a regex. Do you think it's ok?

File: src/charm.py (hunk 2 above)
Comment (shipperizer) [reply src/charm.py:347]: i thought that if something is going thru oathkeeper we want it to be authenticated, the decision endpoint is to take away that logic of checking the header in the client application

i might be completely wrong but, in the case of using the `noop` for `allow` rules we are basically doing a proxy pass, which albeit might be a usecase should probably not be the default one

File: lib/charms/oathkeeper/v0/auth_proxy.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+
+    def restore(self, snapshot: Dict) -> None:
+        """Restore event."""
+        self.relation_id = snapshot["relation_id"]
+
+
+class AuthProxyProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `AuthProxyProvider`."""
+
+    config_changed = EventSource(AuthProxyConfigChangedEvent)
```
Comment (nsklikas) [inline lib/charms/oathkeeper/v0/auth_proxy.py:287]: I would rename this to avoid confusion with the juju `config_changed` event.

```suggestion
    proxy_config_changed = EventSource(AuthProxyConfigChangedEvent)
```

File: lib/charms/oathkeeper/v0/auth_proxy.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+            "type": "array",
+            "default": ["X-User"],
+            "items": {
+                "enum": ALLOWED_HEADERS,
+                "type": "string",
+            },
+        },
+    },
+    "required": ["protected_urls", "allowed_endpoints", "headers"],
+}
```
Comment (nsklikas) [inline lib/charms/oathkeeper/v0/auth_proxy.py:112]: I think that we should start using pydantic to validate the relation databags. We don't have to do it in this PR, but let's create an issue to keep track of it

File: src/charm.py [hunk 7]
```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+        allowed_endpoints: List[str],
+        relation_app_name: str,
+    ) -> Optional[List[Dict]]:
+        """Render access rules from a template."""
+        rules = list()
+
+        for url_index, url in enumerate(protected_urls):
+            if url.startswith("https"):
+                # Replace with regex to match both http and https
+                url = url.replace("https", "<https|http>")
```
Comment (nsklikas) [inline src/charm.py:338]: I think you need to do this regardless of the original URLs scheme.

Traefik may have https enabled, but the app may for some reason place the `http` url in the databag. I think we should parse the URL and replace the scheme with `<https|http>` in all cases.

File: src/charm.py [hunk 8]
```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                # Push the rules to the container
+                filename = f"{self._oathkeeper_access_rules_path}/access-rules-{event.relation_app_name}-{rule_type}.json"
+                self._container.push(filename, str(rules), make_dirs=True)
+                access_rules_locations.append(filename)
+
+        self._set_auth_proxy_relation_peer_data(
+            event.relation_id, dict(access_rules_locations=access_rules_locations)
+        )
+
+        self._push_oathkeeper_config()
```
Comment (nsklikas) [inline src/charm.py:311]: If pushing the file fails, the application will believe that it is protected, but it won't be. I think we should first push the config and then set the peer data.

File: src/charm.py [hunk 9]
```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                    rules.append(allow_rule)
+
+            if rule_type == "deny":
+                if allowed_endpoints:
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        if "/" in endpoint:
+                            # about/app must become a group: (about/app)
+                            endpoint = f"({endpoint})"
```
Comment (nsklikas) [inline src/charm.py:361]: Why is this needed? Perhaps leave a comment about it?

File: src/charm.py [hunk 10]
```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                # Replace with regex to match both http and https
+                url = url.replace("https", "<https|http>")
+
+            if rule_type == "allow":
+                if not allowed_endpoints:
+                    return None
+                for endpoint in allowed_endpoints:
+                    allow_rule = self._rule_template(
+                        rule_id=f"{relation_app_name}:{endpoint}:{url_index}:allow",
+                        url=f"{url}/{endpoint}",
```
Comment (nsklikas) [inline src/charm.py:346]: What if the url ends with `/` will this cause a problem? Should we parse and reconstruct the URLs?

File: src/charm.py [hunk 11]
```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        if "/" in endpoint:
+                            # about/app must become a group: (about/app)
+                            endpoint = f"({endpoint})"
+                        exclude_endpoints.append(endpoint)
+
+                    # Add | alternation
+                    deny_regex = f"{url}/<(?!{'|'.join(exclude_endpoints)}).*>"
```
Comment (nsklikas) [inline src/charm.py:365]: With this regex we catch all endpoints that start with path. 
If someone wanted to allow access to `http://localhost/home`, we would create this rule:
`http://localhost/<(?!home).*>`
BUT if there is a path`http://localhost/home-protected`, which they wanted to protect, it would not be protected.

I think that we should add to all excluded endpoints the prefix `^` and the suffix `$`. So the regex would be:
`^http://localhost/<(?!home$).*>`

Comment (natalian98) [reply src/charm.py:371]: Thanks for catching that, I changed the regex

File: src/charm.py (hunk 8 above)
Comment (natalian98) [reply src/charm.py]: I'm adding the relation's access rules locations to peer data first so that the config gets updated with existing peer data + new dirs rather than overwritten. I put it in a try-except block instead. Is it ok for you?

File: lib/charms/oathkeeper/v0/auth_proxy.py (hunk 6 above)
Comment (natalian98) [reply lib/charms/oathkeeper/v0/auth_proxy.py:112]: Ok, I created an issue

File: src/charm.py (hunk 1 above)
Comment (natalian98) [reply src/charm.py:305]: As agreed during the daily, it will be addressed in a separate PR

**Review by shipperizer — DISMISSED** [review]

File: src/charm.py [hunk 12]
```diff
@@ -213,6 +278,148 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+
+                    rules.append(allow_rule)
+
+            if rule_type == "deny":
+                if allowed_endpoints:
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        endpoint = f"{endpoint}$"
+                        exclude_endpoints.append(endpoint)
```
Comment (nsklikas) [inline src/charm.py:368]: This is a little more pythonic
```suggestion
                    exclude_endpoints = [endpoint + "$" for endpoint in allowed_endpoints]
```

File: src/charm.py [hunk 13]
```diff
@@ -213,6 +278,148 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+            # Replace with regex to match both http and https
+            url = url.replace("https", "<^(https|http)>")
+
+            if rule_type == "allow":
+                if not allowed_endpoints:
+                    return None
+                for endpoint in allowed_endpoints:
+                    allow_rule = self._rule_template(
+                        rule_id=f"{relation_app_name}:{endpoint}:{url_index}:allow",
+                        url=f"{url}/{endpoint}",
```
Comment (nsklikas) [inline src/charm.py:354]:
```suggestion
                        url=f"^{url}/{endpoint}$",
```

File: src/charm.py [hunk 14]
```diff
@@ -213,6 +278,148 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+            if rule_type == "deny":
+                if allowed_endpoints:
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        endpoint = f"{endpoint}$"
+                        exclude_endpoints.append(endpoint)
+
+                    # Add | alternation
+                    deny_regex = f"{url}/<(?!{'|'.join(exclude_endpoints)}).*>"
```
Comment (nsklikas) [inline src/charm.py:371]:
```suggestion
                    deny_regex = f"^{url}/<(?!{'|'.join(exclude_endpoints)}).*>"
```

File: src/charm.py [hunk 15]
```diff
@@ -213,6 +278,148 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
...
+                else:
+                    # Protect all endpoints
+                    deny_regex = f"{url}/<.*>"
```
Comment (nsklikas) [inline src/charm.py:374]:
```suggestion
                    deny_regex = f"^{url}/<.*>"
```

**Review by nsklikas — COMMENTED** [review]
Mostly minor comments, otherwise LGTM.

I have only one question, why do we need to store the file locations in the peer relation? Couldn't we somehow construct the paths from the relation ID? I suppose that once we move to configMaps this logic will change so we don't need to change it.

File: src/charm.py (hunk 13 above)
Comment (natalian98) [reply src/charm.py:354]: `^` is covered a few lines above in `url = url.replace("https", "<^(https|http)>")`. `$` is not needed because the rest of the allowed urls are exact matches, not regex.

File: src/charm.py (hunk 14 above)
Comment (natalian98) [reply src/charm.py:368]: This is covered a few lines above in `url = url.replace("https", "<^(https|http)>")`

File: src/charm.py (hunk 15 above)
Comment (natalian98) [reply src/charm.py:371]: This is covered a few lines above in `url = url.replace("https", "<^(https|http)>")`

**Comment by natalian98** [comment]
> I have only one question, why do we need to store the file locations in the peer relation? Couldn't we somehow construct the paths from the relation ID? I suppose that once we move to configMaps this logic will change so we don't need to change it.

I put the relation id along with its files locations mostly to cater for relation-departed events for both auth-proxy and the forward-auth interface we'll implement next. I think it's better to store this data in a peer relation in case the relation id is unaccessible but we still need it to fire another custom event (update forward-auth), or if a charm got only deny rules created, for example.

**Review by shipperizer — APPROVED** [review]

**Review by nsklikas — APPROVED** [review]
